classes/admin_setting_permitted_directories.php

```diff
@@ -0,0 +1,63 @@
+<?php
+// This file is part of Moodle - https://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_dataflows;
+
+/**
+ * A custom setting for the permitted directories that a dataflow can interact with.
+ *
+ * @package   tool_dataflows
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class admin_setting_permitted_directories extends \admin_setting_configtextarea {
+
+    /**
+     * Validate the setting data as a list of filepaths.
+     *
+     * @param $data array Data to be validated.
```

Comment:
```suggestion
     * @param array $data Data to be validated.
```
The type should come before the variable name. Also is the data always an array? I see at line 42 that it's being trimmed which might indicate that it is in fact a string? Could you please confirm and add the appropriate type hints to the function parameters as well please.

Reply:
thanks